Move gp toolkit ao func tests from tinc

hlinnaka opened the pull request:
I'm not very proud of the hacks needed to force some of these functions to run in segments, rather than the master. We really need a user-accessible syntax for that. We have gp_dist_random() for tables, but AFAIK it cannot be used with functions. But that's a different story. In this particular case, I'm not 100% sure the tests are worth it, so I wouldn't feel bad just removing the hack, and only running these functions in the master.

danielgustafsson commented on src/test/regress/greenplum_schedule:
```diff
@@ -84,6 +84,7 @@ test: wrkloadadmin
 # might not happen if other backends are holding transactions open, preventing
 # vacuum from removing dead tuples.
 test: gp_toolkit
+test: gp_toolkit_ao_funcs
```
Perhaps a comment that concurrent activity will affect output is warranted, to keep this from being reorganized into another group, or more added it to this group?

hlinnaka replied:
It can actually run in parallel, I just wanted to keep it close to the gp_toolkit test.

I added a commit to this PR to rearrange some tests around this.

danielgustafsson replied:
Aha, I interpreted this part of the .out diff as it being dependent on running alone:
```diff
+-- The exact output varies depending on cluster configuration and concurrent
+-- activity in the system
```

ashwinstar commented on src/test/regress/sql/gp_toolkit_ao_funcs.sql:
```diff
@@ -0,0 +1,51 @@
+UPDATE pg_proc set prodataaccess='s' WHERE proname IN
+  ('__gp_aovisimap', '__gp_aovisimap_name', '__gp_aovisimap_entry', '__gp_aovisimap_entry_name');
+
+SELECT * FROM gp_toolkit.__gp_aovisimap('toolkit_ao_test'::regclass) LIMIT 1;
```
I think you can use this trick instead `select gp_toolkit.__gp_aovisimap('toolkit_ao_test'::regclass) from gp_dist_random('gp_id');` gp_id makes it to dispatch to segments.

Not sure why `LIMIT 1` is used ? doesn't that make test output flaky.

hlinnaka commented:
On 02/22/2017 06:04 PM, Ashwin Agrawal wrote:
> ashwinstar commented on this pull request.
>
>
>
>> +SELECT count(*) FROM gp_toolkit.__gp_aovisimap_hidden_info_name('toolkit_ao_test');
> +SELECT * FROM gp_toolkit.__gp_aovisimap_entry('toolkit_ao_test'::regclass);
> +SELECT * FROM gp_toolkit.__gp_aovisimap_entry_name('toolkit_ao_test');
> +SELECT count(*) FROM gp_toolkit.__gp_aoseg_name('toolkit_ao_test');
> +
> +-- These functions don't do anything interesting in the master. We really ought
> +-- to run them in utility mode. To simulate that, temporarily set
> +-- prodataaccess='s' for them, to tell the planner that they must be run on
> +-- all the segments.
> +BEGIN;
> +SET allow_system_table_mods='dml';
> +
> +UPDATE pg_proc set prodataaccess='s' WHERE proname IN
> +  ('__gp_aovisimap', '__gp_aovisimap_name', '__gp_aovisimap_entry', '__gp_aovisimap_entry_name');
> +
> +SELECT * FROM gp_toolkit.__gp_aovisimap('toolkit_ao_test'::regclass) LIMIT 1;
>
> I think you can use this trick instead `select gp_toolkit.__gp_aovisimap('toolkit_ao_test'::regclass) from gp_dist_random('gp_id');` gp_id makes it to dispatch to segments.

Thanks! Changed to do that instead.

> Not sure why `LIMIT 1` is used ? doesn't that make test output flaky.

Ah, right, that was very subtle. The point was to make the test output 
the same, regardless of the number of segments in the cluster. The test 
INSERTed some rows, then UPDATEd one, and then DELETEd one. The test 
output would depend on whether the UPDATEd and DELETEd rows were on the 
same segment, or different segment:

        tid        | segno | row_num
------------------+-------+---------
  (33554432,32769) |     1 |       1
  (33554432,32770) |     1 |       2
(2 rows)

vs

        tid        | segno | row_num
------------------+-------+---------
  (33554432,32769) |     1 |       1
  (33554432,32769) |     1 |       1
(2 rows)


With the LIMIT 1, it would always return just the first row, which is 
the same in both cases.

But but that really is too subtle. I've modified the test so that all 
the rows are on the same segment, so that you always get the 1st output 
above, and removed the LIMIT 1.

Updated the PR, please have another look.

- Heikki

hlinnaka commented:
Pushed, thanks for the review!
